### ISSUE TITLE
AWSS3TransferUtility uploadFileUsingMultiPart in background returns HTTP load failed (error code: -1001

### ISSUE THREAD

**Issue by rivkiG** [aws-amplify/aws-sdk-ios#1088]
hi, I am working on an app that uploads also in the background in worked fine until I updated the WAS framework to 2.6.31 and now I get lots of this error:
Task <D223536B-8BBD-4EEA-96D2-AA668C3B8592>.<1049> load failed with error Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=-2102, NSUnderlyingError=0x281451530 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <D223536B-8BBD-4EEA-96D2-AA668C3B8592>.<1049>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <D223536B-8BBD-4EEA-96D2-AA668C3B8592>.<1049>"
), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/CameraRoll/102APPLE-IMG_2972.JPG?uploads=, NSErrorFailingURLKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/CameraRoll/102APPLE-IMG_2972.JPG?uploads=, _kCFStreamErrorDomainKey=4} [-1001]

what could be the problem?
thank you

**Comment by cbommas** [comment]
@rivkiG 

Sorry to hear that you are running into issues. Can you post a code snippet from your side - especially around how you are instantiating the TransferUtility object?

**Comment by rivkiG** [comment]
thank you for your answer,  
this is my AWSS3TransferUtility object:
 AWSS3TransferUtility *transferUtility = [AWSS3TransferUtility S3TransferUtilityForKey:kCellularConfigurationName];
    [ transferUtility enumerateToAssignBlocksForUploadTask:^(AWSS3TransferUtilityUploadTask * _Nonnull uploadTask, AWSS3TransferUtilityProgressBlock  _Nullable __autoreleasing * _Nullable uploadProgressBlockReference, AWSS3TransferUtilityUploadCompletionHandlerBlock  _Nullable __autoreleasing * _Nullable completionHandlerReference) {
        
        NSLog(@"%lu", (unsigned long)uploadTask.taskIdentifier);
        
//        *uploadProgressBlockReference = _progressBlock;
//        *completionHandlerReference   = _uploadCompletionHandler;
        
    } multiPartUploadBlocksAssigner:^(AWSS3TransferUtilityMultiPartUploadTask * _Nonnull multiPartUploadTask, AWSS3TransferUtilityMultiPartProgressBlock  _Nullable __autoreleasing * _Nullable multiPartUploadProgressBlockReference, AWSS3TransferUtilityMultiPartUploadCompletionHandlerBlock  _Nullable __autoreleasing * _Nullable completionHandlerReference) {
        
        NSLog(@"%lu", (unsigned long)multiPartUploadTask.transferID);
        
        *multiPartUploadProgressBlockReference = _multiPartProgressBlock;
        *completionHandlerReference   = _multiPartUploadCompletionHandler;
    } downloadBlocksAssigner:^(AWSS3TransferUtilityDownloadTask * _Nonnull downloadTask, AWSS3TransferUtilityProgressBlock  _Nullable __autoreleasing * _Nullable downloadProgressBlockReference, AWSS3TransferUtilityDownloadCompletionHandlerBlock  _Nullable __autoreleasing * _Nullable completionHandlerReference)
     {
         
     }];
this is the AWSS3TransferUtilityConfiguration object:
AWSS3TransferUtilityConfiguration * tfuConfig = [[AWSS3TransferUtilityConfiguration alloc] init];
    tfuConfig.retryLimit = 10;
    tfuConfig.multiPartConcurrencyLimit = @1;
    tfuConfig.timeoutIntervalForResource = 50*10;

and this is my actual uploadFileUsingMultiPart call:
 AWSS3TransferUtility *transferUtility = [AWSS3TransferUtility S3TransferUtilityForKey:[self allowCellular] ? kCellularConfigurationName : kNoCellularConfigurationName];
    [[transferUtility uploadFileUsingMultiPart:fileURL bucket:_bucket key:file.targetPath contentType:fileContentTypeStr expression:expression completionHandler:_multiPartUploadCompletionHandler] continueWithBlock:^id(AWSTask *task) {
        if (task.error ) {
            ErrorStatus * errorStatus;
            if ( task.error ) {
                NSLog(@"Upload Error: %@ for file %@", task.error, fileURL);
                errorStatus = [ErrorStatus errorWithDomain:MediaMagnetErrorDomain code:task.error.code userInfo:nil];
            }
            else {
                NSLog(@"Upload Exception: %@ for file %@", task.error, fileURL);
                errorStatus = [ErrorStatus errorWithDomain:MediaMagnetErrorDomain code:NoAccessToFile userInfo:nil];
            }
            [_callback transferFinished:Upload fileName:file.targetPath fileType:file.fileType errorStatus:errorStatus];
        }
        return nil;
    }];

thank you for your help

**Comment by rivkiG** [comment]
its a mistake it needed to be 30000: 50*60
could that be the problem?

**Comment by cbommas** [comment]
That certainly is a contributing factor. Are you able to try it out with the longer timeout and see if that helps?

**Comment by rivkiG** [comment]
yes im doing it now and I will update you

**Comment by rivkiG** [comment]
I still get this error:
Task <39B990DD-A36B-4AB8-85A1-073320A39404>.<563> load failed with error Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=-2102, NSUnderlyingError=0x28062ae20 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <39B990DD-A36B-4AB8-85A1-073320A39404>.<563>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <39B990DD-A36B-4AB8-85A1-073320A39404>.<563>"
), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/photos%201000/100SYNCD-IMG_0031.JPG?uploads=, NSErrorFailingURLKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/photos%201000/100SYNCD-IMG_0031.JPG?uploads=, _kCFStreamErrorDomainKey=4} [-1001]

**Comment by cbommas** [comment]
Some more questions
1) How big is the file you are trying to transfer and how long does it typically take?
2) Are you running this on a device or Simulator?
3) Are you on WiFi or Cellular?

**Comment by rivkiG** [comment]
1. 1.5 MB it takes a few seconds 
2. I'm running it on a device.
3. I'm on wifi.

it works well when the app is in the front ground but in the background after Background expired this error starts.

**Comment by cbommas** [comment]
@rivkiG 

A suggestion and a question.
1) Suggestion: have you implemented the `handleEventsForBackgroundURLSession` method in you `AppDelegate` as mentioned here https://docs.aws.amazon.com/aws-mobile/latest/developerguide/how-to-transfer-files-with-transfer-utility.html? 
2) Question: What do you mean by "Background Expired" - Is the app being terminated in the background by the system or are you force closing it?

**Comment by rivkiG** [comment]
yes, I call handleEventsForBackgroundURLSession 
and the  "Background Expired"  - It's being terminated in the background by the system. but i see that it happens right when I go to background it has nothing to do with the background system - my mistake.
I see also that I get -1005 error:
 Task <27833976-A1CA-4049-A307-08A8C8A039D3>.<3129> load failed with error Error Domain=NSURLErrorDomain Code=-1005 "The network connection was lost." UserInfo={_kCFStreamErrorCodeKey=-4, NSUnderlyingError=0x282a0e880 {Error Domain=kCFErrorDomainCFNetwork Code=-1005 "(null)" UserInfo={NSErrorPeerAddressKey=<CFData 0x281956800 [0x1ff0e15e0]>{length = 16, capacity = 16, bytes = 0x100201bb34d8e1d30000000000000000}, _kCFStreamErrorCodeKey=-4, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <27833976-A1CA-4049-A307-08A8C8A039D3>.<3129>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <27833976-A1CA-4049-A307-08A8C8A039D3>.<3129>"
), NSLocalizedDescription=The network connection was lost., NSErrorFailingURLStringKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/CameraRoll/103APPLE-IMG_3650.JPG?uploadId=Vf5i1LAePLg3uDQSuCxr77cM6pwp_ZNbqD5Sd1bjzSrKI9oLqNEGyC3zXiw73pdh28RvVSo_sSzWmc3dmdvMS4N12rPAjfR5Ef93D8bXT3o7E3rSrU9qWAfYfODmDNDISLAZk40m.cImVtxbhfGvF5LtOfF1VXr9Dg2HsttYBTE-, NSErrorFailingURLKey=https://s3.dualstack.us-east-1.amazonaws.com/hyperlync-cm-production/ef7f3161-c283-4ee2-b21e-c1ea8f8dc68d/8cb94ec8-d4f5-44f3-b625-8a6690e04986/Photos/CameraRoll/103APPLE-IMG_3650.JPG?uploadId=Vf5i1LAePLg3uDQSuCxr77cM6pwp_ZNbqD5Sd1bjzSrKI9oLqNEGyC3zXiw73pdh28RvVSo_sSzWmc3dmdvMS4N12rPAjfR5Ef93D8bXT3o7E3rSrU9qWAfYfODmDNDISLAZk40m.cImVtxbhfGvF5LtOfF1VXr9Dg2HsttYBTE-, _kCFStreamErrorDomainKey=4} [-1005]

there is something with the connection that was changed maybe?
and another question do I have to put a timeout for a file? if there is a have file of 50 MB and slow wifi? can it be open? no end?
I mean: timeoutIntervalForResource?
thank you

**Comment by cbommas** [comment]
@rivkiG 

When you use Cognito for authentication, you are unfortunately bound by the 50 minute time limit for `timeoutIntervalForResource`.

Here are some options that might help you get around the situation that you are encountering
* increase the value for `multiPartConcurrencyLimit` from 1 to a higher value, say 5. This controls how many parts get uploaded in parallel. So for a 50 MB file, you will have 5 parts being uploaded in parallel, which should speed things up. You can adjust this up or down to meet your specific circumstance. This would be the solution that I'd recommend.
* another option is to use static credentials that don't expire within an hour. With this, you can set the `timeoutIntervalForResource` to a much higher value ( up to a maximum of 7 days). However, this option is not great from a security standpoint as you will have to figure out a secure way of using static credentials in your app, and as such not recommended. See below for a code snippet showing how to instantiate the TransferUtility with static credentials 

        let staticCredentialProvier = AWSStaticCredentialsProvider(accessKey: "ACCESSKEY", secretKey: "SECRET KEY")
        let serviceConfig = AWSServiceConfiguration(region: .USEast1, credentialsProvider: staticCredentialProvider)
        AWSS3TransferUtility.register(with: serviceConfig!, forKey: "static")

**Comment by rivkiG** [comment]
thank you for your ideas
do you have a solution may be for the connection issue?

**Comment by rivkiG** [comment]
hi
I get another crash maybe you will have an answer for it.
if I do a big backup after a while I get this error: 
'Task created in a session that has been invalidated" and it crashes the app I saw that it crashes in this point: 
in the AWSURLSessionManager.m file in the method: didCompleteWithError of NSURLSessionTaskDelegate.
in the line:  [[[AWSTask taskWithResult:nil] continueWithSuccessBlock:^id(AWSTask *task) { number 216
and do you have a solution to the connection issue?
thank you

**Comment by cbommas** [comment]
@rivkiG 
For the connection error, I think it may be related to a change we made to create NSURLSession tasks in a "just in time" fashion for multi-part uploads as opposed to all at once. This is resulting in some of the tasks being created in the background and having a low priority assigned to them by the OS. 

I am working to get this reverted back and this may help with reducing/resolving the timeout issues. I am curious though - did increasing the `multiPartConcurrencyLimit` help?

For the crash, I have posted a response on https://github.com/aws-amplify/aws-sdk-ios/issues/1135 and lets track it there to keep these two issues separate.

**Comment by rivkiG** [comment]
thank you for the answer the multiPartConcurrencyLimit didn't help it even though it's much better.
if you can please let me know  about any updates on this

**Comment by minbi** [comment]
Hi @rivkiG,

We have just released version 2.8.2 of the SDK. Please check if this fixes the issue for you.

**Comment by cbommas** [comment]
@rivkiG 
Version 2.8.2 creates NSURLSession tasks up front and this should help increase the speed of transfers and reduce the timeouts you are encountering.
